crates/subspace-block-relay/src/lib.rs

```diff
@@ -0,0 +1,153 @@
+//! Block relay implementation.
+//!
+//! The components in the system:
+//! 1. Relay users like consensus, execution. They implement the use case
+//!    specific logic that drives the relay protocol. This has a client
+//!    side stub, and a server side task to process incoming requests.
+//! 2. Relay protocol that is agnostic to the relay user. The protocol
+//!    is abstracted to be reused for different use cases. The protocol
+//!    also has corresponding client/server side components.
+//! 3. Protocol backend: relay user specific abstraction used by the relay
+//!    protocol to populate the protocol messages
+//!
+//! Nodes advertise/exchange DownloadUnits with each other. DownloadUnit has
+//! two parts:
+//! - ProtocolUnits: the part fetched by the relay protocol. This is bulk of
+//!   the data transfer that we would like to optimize
+//! - Rest of the download unit, handled directly by the relay user
+//!
+//! Examples:
+//! 1. Consensus
+//!    DownloadUnit = Block, ProtocolUnit = extrinsics
+//!    The extrinsics are handled by the protocol, remaining block
+//!    fields are directly filled by the caller. The protocol backend
+//!    helps fetch blocks/transactions from the substrate backend
+//! 2. Execution
+//!    TODO
+//! 3. Other possible use cases (e.g) reconcile/sync the transaction pool
+//!    between two nodes. In this case, DownloadUnit = transaction pool,
+//!    ProtocolUnit = transaction
+//!
+//! The download has two phases:
+//! -  Initial request/response
+//!    Ideally, download of all the protocol units in the download unit should
+//!    be completed during this phase
+//! -  Reconcile phase
+//!    If the initial phase could not complete the download, additional
+//!    request/response messages are initiated by the protocol to fetch the
+//!    protocol units
+//!
+
+#![feature(const_option)]
+
+use crate::utils::RelayError;
+use async_trait::async_trait;
+use codec::{Decode, Encode};
+use futures::channel::oneshot::Canceled;
+use sc_network::RequestFailure;
+use std::sync::Arc;
+use std::time::Duration;
+
+mod consensus;
+mod protocol;
+mod utils;
+
+pub use crate::consensus::build_consensus_relay;
+pub use crate::utils::NetworkWrapper;
+
+pub(crate) const LOG_TARGET: &str = "block_relay";
+
+pub(crate) type EncoderFn<T> = Box<dyn Fn(T) -> Vec<u8> + Send>;
+
+/// The downloaded entry and meta info
+pub(crate) struct DownloadResult<DownloadUnitId> {
+    /// Downloaded unit Id
+    download_unit_id: DownloadUnitId,
+
+    /// Downloaded entry
+    download_unit: Vec<u8>,
```

Comment:
Can this also be a generic?

Reply:
It is much easier to trace back generic than place where `Vec<u8>` turns into something more meaningful.